riscv/triggers.cc

```diff
@@ -17,6 +17,51 @@ action_t trigger_t::legalize_action(reg_t val) const noexcept {
   return (val > ACTION_MAXVAL || (val == ACTION_DEBUG_MODE && get_dmode() == 0)) ? ACTION_DEBUG_EXCEPTION : (action_t)val;
 }
 
+unsigned trigger_t::legalize_mhselect(reg_t val, bool h_enabled) const noexcept {
+  switch (val) {
+    case MHSELECT_IGNORE:
+    case MHSELECT_MCONTEXT:
+      return val;
+    case 1:
+    case 2:
+      if (!h_enabled)
+        return MHSELECT_IGNORE;
+      return val;
+    case 3: return MHSELECT_IGNORE;
+    case 5:
+    case 6:
+      if (!h_enabled)
+        return MHSELECT_MCONTEXT;
+      return val;
+    case 7: return MHSELECT_IGNORE;
+    default: assert(false);
+  }
+}
+
+reg_t trigger_t::tdata3_read(const processor_t * const proc) const noexcept {
+  auto xlen = proc->get_xlen();
+  reg_t tdata3 = 0;
+  tdata3 = set_field(tdata3, CSR_TEXTRA_MHVALUE(xlen), mhvalue);
+  tdata3 = set_field(tdata3, CSR_TEXTRA_MHSELECT(xlen), mhselect);
+  tdata3 = set_field(tdata3, CSR_TEXTRA_SBYTEMASK(xlen), sbytemask);
+  tdata3 = set_field(tdata3, CSR_TEXTRA_SVALUE(xlen), svalue);
+  tdata3 = set_field(tdata3, CSR_TEXTRA_SSELECT(xlen), sselect);
+  return tdata3;
+}
+
+void trigger_t::tdata3_write(processor_t * const proc, const reg_t val) noexcept {
+  auto xlen = proc->get_xlen();
+  mhvalue = get_field(val, CSR_TEXTRA_MHVALUE(xlen));
+  mhselect = legalize_mhselect(get_field(val, CSR_TEXTRA_MHSELECT(xlen)), proc->extension_enabled('H'));
+  sbytemask = get_field(val, CSR_TEXTRA_SBYTEMASK(xlen));
+  svalue = get_field(val, CSR_TEXTRA_SVALUE(xlen));
+  sselect = get_field(val, CSR_TEXTRA_SSELECT(xlen));
+  if (!proc->extension_enabled_const('S'))
+    svalue = 0;
+  if (!proc->extension_enabled_const('S') || sselect > SSELECT_MAXVAL)
+    sselect = SSELECT_IGNORE;
+}
```

Comment:
I would prefer to set these two in one expression each, instead of setting them and then modifying them in a later line of code, since a future maintainer could easily miss that later modification.